lib/LongVectorLoweringPass.cpp

```diff
@@ -1382,19 +1419,36 @@ Type *clspv::LongVectorLoweringPass::getEquivalentTypeImpl(Type *Ty) {
     LLVMContext &Ctx = StructTy->getContainedType(0)->getContext();
     SmallVector<Type *, 16> Types;
     bool RequiredLowering = false;
+    bool Packed = StructTy->isPacked();
     for (unsigned i = 0; i < Arity; ++i) {
       Type *CTy = StructTy->getContainedType(i);
       auto *EquivalentTy = getEquivalentType(CTy);
       if (EquivalentTy != nullptr) {
         Types.push_back(EquivalentTy);
         RequiredLowering = true;
+
+        auto InitialOff = DL->getStructLayout(StructTy)->getElementOffset(i);
+        auto NewOff = DL->getStructLayout(StructType::get(Ctx, Types, Packed))
+                          ->getElementOffset(Types.size() - 1);
+        if (InitialOff != NewOff) {
+          Types.pop_back();
+          Types.push_back(
+              ArrayType::get(Ty->getInt8Ty(Ctx), InitialOff - NewOff));
+          Types.push_back(EquivalentTy);
+        }
       } else {
         Types.push_back(CTy);
       }
     }
 
     if (RequiredLowering) {
-      return StructType::get(Ctx, Types, false);
+      auto InitialSize = DL->getTypeAllocSize(StructTy);
+      auto NewSize = DL->getTypeAllocSize(StructType::get(Ctx, Types, Packed));
+      if (InitialSize != NewSize) {
+        Types.push_back(
+            ArrayType::get(Ty->getInt8Ty(Ctx), InitialSize - NewSize));
```

Comment:
This seems to add an Int8 requirement that might not otherwise exist. I suggest using an i32 array if the size is evenly divisible by 4.

Reply:
Done